Added cctp-bridge overview.md (#393) (#411)
wormhole-foundation/wormhole-docs: products/cctp-bridge/overview.md

categories: Transfer
---

# CCTP with Wormhole Overview 

The integration of [Circle's Cross-Chain Transfer Protocol (CCTP)](https://www.circle.com/cross-chain-transfer-protocol){target=\_blank} with the Wormhole messaging protocol creates a robust system for securely and efficiently transferring native USDC across different blockchain networks while enabling more complex multichain interactions. This combination streamlines the movement of stablecoins, reduces risk, and unlocks new possibilities for decentralized applications.

## Key Features

- **Secure native USDC transfers** - at its core, CCTP provides a "burn-and-mint" mechanism for transferring native USDC. This eliminates the need for wrapped assets and the associated risks of intermediary bridges
- **Atomic execution** - by combining CCTP and Wormhole, the transfer of USDC and the execution of accompanying instructions on the destination chain can occur as a single, atomic transaction
- **Automated relaying** - eliminates the need for users to redeem USDC transfers themselves
- **Gas payment on the destination chain** - allows users to transfer USDC without needing to pay gas on the destination chain

> - **Gas drop off** - enables users to convert a portion of USDC into the destination chain's gas token upon a successful transfer

Reviewer: here it's first saying you don't need to pay gas on the destination chain and then it's saying you can convert USDC to pay gas on the destination chain, they can't both be true at the same time 
i see we had this on the website but it's confusing, can you look into it which one it is

Author: From what I found, based on automatic or manual, can change if gas is taken from the source and destination or if its just the destination. there is no minimum amount for bridging via CCTP if the user covers the gas fees on both the source and destination chains. However, if the transfer is automatically relayed, a minimum amount is required to cover relay fees on the destination chain.

Then, the 'Gas drop off' feature seems to be a mechanism that helps make this possible. It allows some of the USDC you're transferring to be converted into gas on the destination chain.

Reviewer: okay then we should give context that this happens when its automatic or manual

> - **Enhanced composability** - developers can build more sophisticated cross-chain applications by sending additional data alongside the 

Reviewer: alongside the what

## How It Works

This section outlines the end-to-end flow for transferring native USDC across chains using CCTP while optionally triggering an action on the destination chain. Circle and Wormhole coordinate each step to ensure a secure, verifiable transfer and execution process.

1. **Initiation on source chain** - a user on the source chain initiates a USDC transfer and specifies the action on the destination chain
2. **CCTP burn and package** - Wormhole automates the transfer by packaging the USDC burn request and destination chain action instructions into a single message
3. **Circle attestation** -  Circle's attestation service confirms the USDC burn on the source chain and issues a signature that is packaged together with the Wormhole message
4. **Guardians verify and sign** - Wormhole's Guardians observe and sign the package achieving consensus on validity 

> 5. **Automated relay** - the Relayers then automatically push the package to the destination chain without manual interaction

Reviewer: ```suggestion
5. **Automated relay** - the Relayers then automatically deliver the package to the destination chain without manual interaction
```

6. **Mint and verify** - Circle's attestation in Wormhole message triggers the minting of native USDC on the destination chain after verification
7. **Execute action** - any action on the destination chain executes if included in Wormhole message, potentially using new USDC

> 

Reviewer: ```suggestion
```

    I --> J
    J -- Submit VAA & Attestation --> K
    K -- Verify Wormhole VAA --> L
    K -- Call Circle CCTP Contract --> M
    K -- (Optional) Execute Payload Logic --> N
    M -- Mint Native USDC --> O
    N --> O
    style B fill:#F5F5F5,stroke:#333,stroke-width:1px,color:#000
    style A fill:#FFFFFF,stroke:#333,stroke-width:2px,color:#000
    style C fill:#D9D9D9,stroke:#333,stroke-width:2px,color:#000
    style D fill:#ADD8E6,stroke:#333,stroke-width:1px,color:#000
    style E fill:#FFFFCC,stroke:#333,stroke-width:1px,color:#000
    style F fill:#F0F8FF,stroke:#333,stroke-width:2px,color:#000
    style G fill:#E6E6FA,stroke:#333,stroke-width:2px,color:#000
    style H fill:#E0FFFF,stroke:#333,stroke-width:1px,color:#000
    style I fill:#FAFAD2,stroke:#333,stroke-width:1px,color:#000
    style J fill:#FFDAB9,stroke:#333,stroke-width:2px,color:#000
    style K fill:#D9D9D9,stroke:#333,stroke-width:2px,color:#000
    style L fill:#ADD8E6,stroke:#333,stroke-width:1px,color:#000
    style M fill:#ADD8E6,stroke:#333,stroke-width:1px,color:#000
    style N fill:#F5F5F5,stroke:#333,stroke-width:1px,color:#000
    style O fill:#FFFFFF,stroke:#333,stroke-width:2px,color:#000

> ```

Reviewer: this diagram is way too big and it's not using the mkdocs styling 
remove all those styles fromt the diagram, we deal with styling in mkdocs, the mkdocs staging-new-ia should have the right styles already

!!! note 
    Wormhole supports all CCTP chains, however, Circle currently supports a few that you can find listed in [Circles supported domains](https://developers.circle.com/stablecoins/supported-domains).

> 

Reviewer: ```suggestion
```

- **Gas payment** - covering destination gas in automated vs. manual transfers:
    - **Automated** - users often don't need destination gas tokens upfront, relayers cover these gas costs, reimbursed via gas drop-off or initial fees
    - **Manual** - users pay destination gas directly, the protocol may offer post-claim USDC-to-gas conversion

## How It Works

This section outlines the end-to-end flow for transferring native USDC across chains using CCTP while optionally triggering an action on the destination chain. Circle and Wormhole coordinate each step to ensure a secure, verifiable transfer and execution process.

1. **Initiation on source chain** - a user on the source chain initiates a USDC transfer and specifies the action on the destination chain
2. **CCTP burn and package** - Wormhole automates the transfer by packaging the USDC burn request and the destination chain action instructions into a single message
3. **Circle attestation** -  Circle's attestation service confirms the USDC burn on the source chain and issues a signature that is packaged together with the Wormhole message

> 4. **Guardians verify and sign** - Wormhole's Guardians observe and sign the package, achieving consensus on validity 

Reviewer 2: ```suggestion
4. **Guardians verify and sign** - Wormhole's [Guardians](/docs/protocol/infrastructure/guardians/){target=\_blank} observe and sign the package, achieving consensus on validity 
```

> 5. **Automated relay** - the Relayers then automatically deliver the package to the destination chain without manual interaction

Reviewer 2: ```suggestion
5. **Automated relay** - the [relayers](/docs/protocol/infrastructure/relayer/){target=\_blank} then automatically deliver the package to the destination chain
```

- **Gas drop off** - enables users to convert a portion of USDC into the destination chain's gas token upon a successful transfer
- **Enhanced composability** - developers can build more sophisticated cross-chain applications by sending additional data alongside the transfer

## How It Works

This section outlines the end-to-end flow for transferring native USDC across chains using CCTP while optionally triggering an action on the destination chain. Circle and Wormhole coordinate each step to ensure a secure, verifiable transfer and execution process.

1. **Initiation on source chain** - a user on the source chain initiates a USDC transfer and specifies the action on the destination chain
2. **CCTP burn and package** - Wormhole automates the transfer by packaging the USDC burn request and the destination chain action instructions into a single message
3. **Circle attestation** -  Circle's attestation service confirms the USDC burn on the source chain and issues a signature that is packaged together with the Wormhole message

> 4. **Guardians verify and sign** - Wormhole's Guardians observe and sign the package achieving consensus on validity 

Reviewer 2: ```suggestion
4. **Guardians verify and sign** - Wormhole's Guardians observe and sign the package, achieving consensus on validity 
```

5. **Automated relay** - the Relayers then automatically deliver the package to the destination chain without manual interaction
6. **Mint and verify** - Circle's attestation in Wormhole message triggers the minting of native USDC on the destination chain after verification

> 7. **Execute action** - any action on the destination chain executes if included in Wormhole message, potentially using new USDC

Reviewer 2: ```suggestion
7. **Execute action** - any action on the destination chain executes if included in the message, potentially using the new USDC
```

> ```mermaid

Reviewer 2: This diagram covers more than what the steps above it cover. Maybe we can move this to flow of a transfer page (eventually), and create a more simplified one here

Asked ChatGPT to create a diagram based on the set of steps above and this is what it returned:

```mermaid
sequenceDiagram
    participant User
    participant SourceChain as Source Chain
    participant Circle
    participant Guardians as Wormhole Guardians
    participant Relayers
    participant DestinationChain as Destination Chain

    User->>SourceChain: Initiate USDC transfer <br> with destination action
    SourceChain->>Circle: Burn USDC and package <br> transfer + action
    Circle->>Circle: Attest burn and issue signature
    Circle->>Guardians: Package message with <br> Circle signature
    Guardians->>Guardians: Verify & sign package
    Relayers->>Guardians: Pick up signed package
    Relayers->>DestinationChain: Relay package <br> automatically
    DestinationChain->>Circle: Verify Circle attestation
    DestinationChain->>DestinationChain: Mint native USDC
    DestinationChain->>DestinationChain: Execute destination action (if any)
```

Code:

```
```mermaid
sequenceDiagram
    participant User
    participant SourceChain as Source Chain
    participant Circle
    participant Guardians as Wormhole Guardians
    participant Relayers
    participant DestinationChain as Destination Chain

    User->>SourceChain: Initiate USDC transfer <br> with destination action
    SourceChain->>Circle: Burn USDC and package <br> transfer + action
    Circle->>Circle: Attest burn and issue signature
    Circle->>Guardians: Package message with <br> Circle signature
    Guardians->>Guardians: Verify & sign package
    Relayers->>Guardians: Pick up signed package
    Relayers->>DestinationChain: Relay package <br> automatically
    DestinationChain->>Circle: Verify Circle attestation
    DestinationChain->>DestinationChain: Mint native USDC
    DestinationChain->>DestinationChain: Execute destination action (if any)
/```
```

⚠️  **I did not check the diagram for accuracy**

---
config:
  layout: fixed
---
flowchart TD
 subgraph subGraph0["Source Chain"]
        B("Call Wormhole CCTP Integration Contract")
        A["User/dApp"]
        D["Circle CCTP Contract"]
        E("Source Chain Transaction Logs")
  end
 subgraph subGraph1["Wormhole and Circle Network"]
        F["Wormhole Guardian Network"]
        G["Circle Attestation Service"]
        H["Wormhole VAA"]
        I["Signed Circle Attestation"]
        J("Relayer")
  end
 subgraph subGraph2["Target Chain"]
        K["Submit VAA & Attestation"] 
        L["Wormhole Core Contract"]
        M["Circle CCTP Contract"]
        N["Target dApp/Contract"]
        O["Recipient Wallet/Contract"]
  end
    A --> B
    B -- Burn USDC --> D
    B -- Emit Wormhole Message --> E
    E --> F
    D --> G
    F -- Sign VAA --> H
    G -- Sign Attestation --> I
    H --> J
    I --> J
    J --> K 
    K -- Verify Wormhole VAA --> L
    K -- Call Circle CCTP Contract --> M
    K -- (Optional) Execute Payload Logic --> N
    M -- Mint Native USDC --> O
    N --> O
```
!!! note 

>     Wormhole supports all CCTP chains, however, Circle currently supports a few that you can find listed in [Circles supported domains](https://developers.circle.com/stablecoins/supported-domains).

Reviewer 2: It doesn't sound right to say "CCTP chains". I also took some time to look into this statement to see if it's still true. The above statement makes it sound like Wormhole supports more chains than what Circle supports. But this is not true if you compare the supported chains on [Wormhole](https://papermoonio.github.io/wormhole-mkdocs/products/reference/contract-addresses/#cctp) and [Circle](https://developers.circle.com/stablecoins/supported-domains). 

So I would just generalize the statement to: "For a cross-chain transfer to be successful, both the source and destination chains must be among those supported by Circle's CCTP."

@ilariae can you please confirm this?

Reviewer: yeah generalize statement is good 👍
maybe its more right to say - the chains must support Circle's CCTP -

## Use Cases

Integrating Wormhole's messaging with CCTP enables the secure transfer of native USDC across blockchains, unlocking key cross-chain use cases, which include:

- **USDC Payments Across Chains**
    - [**CCTP**](/docs/products/cctp-bridge/get-started/) – transfer native USDC using Circle’s burn-and-mint protocol
    - [**Wormhole TypeScript SDK**](/docs/tools/typescript-sdk/sdk-reference/) – automate attestation delivery and gas handling
    - [**Connect**](/docs/products/connect/overview/) – embed multichain USDC transfers directly in your app

- **USDC-Powered Multichain Settlement**
    - [**Settlement**](/docs/products/settlement/overview/) – use the Liquidity Layer to settle intents with native USDC
    - [**Wormhole TypeScript SDK**](/docs/tools/typescript-sdk/sdk-reference/) – initiate transfers, discover routes, and execute swaps seamlessly

## Next Steps

Now that you're familiar with CCTP, here is a list of resources for more hands-on practice:

> - [**Get started with CCTP Bridge**](Todo) - perform a multochain USDC transfer from Avalanche to Sepolia using Wormhole's TypeScript SDK and Circle's CCTP
> - [**Complete USDC Transfer Flow**](Todo) -  execute a USDC cross-chain transfers using Wormhole SDK and Circle's CCTP, covering manual, automatic, and partial transfer recovery
> - [**Checkout Circle's CCTP Docs**](https://developers.circle.com/stablecoins/cctp-getting-started) - learn more about Circle's cross chain transfer protocol in their documentation

Reviewer 2: ```suggestion
- [**Get started with CCTP Bridge**](/docs/products/cctp-bridge/get-started/) - perform a multichain USDC transfer from Avalanche to Sepolia using Wormhole's TypeScript SDK and Circle's CCTP
- [**Complete USDC Transfer Flow**](Todo) - execute a USDC cross-chain transfer using Wormhole SDK and Circle's CCTP, covering manual, automatic, and partial transfer recovery
- [**Checkout Circle's CCTP Docs**](https://developers.circle.com/stablecoins/cctp-getting-started) - learn more about Circle's cross-chain transfer protocol in their documentation
```

1. **Alice initiates a transfer on Ethereum**: She submits a request to the Circle Bridge to send 100 USDC to Avalanche. If desired, she could include optional payload data.

2. **Tokens are taken into custody and burned**: The Circle Bridge takes custody of Alice's USDC and initiates a burn using Circle's CCTP, triggering an off-chain attestation process.

3. **A Wormhole message is published**: The transfer metadata is emitted as a Wormhole message. [Guardians](/docs/protocol/infrastructure/guardians/){target=\_blank} validate and sign it to produce a [Verifiable Action Approval (VAA)](/docs/protocol/infrastructure/vaas/){target=\_blank}.

4. **A relayer automatically processes the messages**: Once the VAA and Circle attestation are available, a relayer submits them to the Circle Bridge on Avalanche.

5. **Tokens are minted**: The Circle Bridge verifies both proofs and mints 100 USDC to Alice using Circle's CCTP. If a payload is included, it can be executed atomically.

> 

Reviewer: nice!

## Key Features

- **Secure native USDC transfers** - at its core, CCTP provides a "burn-and-mint" mechanism for transferring native USDC. This eliminates the need for wrapped assets and the associated risks of intermediary bridges
- **Atomic execution** - by combining CCTP and Wormhole, the transfer of USDC and the execution of accompanying instructions on the destination chain can occur as a single atomic transaction
- **Automated relaying** - eliminates the need for users to redeem USDC transfers themselves
- **Enhanced composability** - developers can build more sophisticated cross-chain applications by sending additional data alongside the transfer
- **Gas drop off** - enables users to convert a portion of USDC into the destination chain's gas token upon a successful transfer
- **Gas payment** - covering destination gas in automated vs. manual transfers:
    - **Automated** - users often don't need destination gas tokens upfront, relayers cover these gas costs, reimbursed via gas drop-off or initial fees
    - **Manual** - users pay destination gas directly, the protocol may offer post-claim USDC-to-gas conversion

> ## How It Works

Reviewer 2: Could you please provide me with links in the code or something that can verify this flow?

Reviewer 2: I'm most curious about the flow from our CCTP contracts to Circle's attestation service and back to Wormhole

Author: I used these pages https://wormhole.com/docs/build/transfers/cctp/#circles-cctp-contracts  and https://wormhole.com/docs/build/toolkit/typescript-sdk/sdk-layout/#vaas-and-layouts to piece it together, i could have made a mistake somewhere tho

Reviewer 2: Ok, I'll dig this up in the actual code to verify. Thank you

> 1. **Initiation on source chain** - a user on the source chain initiates a USDC transfer and specifies the action on the destination chain

Reviewer 2: ```suggestion
1. **Initiation on source chain** - a user on the source chain initiates a USDC transfer and specifies the action to be executed on the destination chain
```

> 2. **CCTP burn and package** - Wormhole automates the transfer by packaging the USDC burn request and the destination chain action instructions into a single message

Reviewer 2: ```suggestion
2. **Package burn and payload instructions** - Wormhole automates the transfer by packaging the USDC burn request and the destination chain action instructions into a single message
```

6. **Mint and verify** - Circle's attestation in Wormhole message triggers the minting of native USDC on the destination chain after verification
7. **Execute action** - any action on the destination chain executes if included in the message, potentially using the new USDC

```mermaid
sequenceDiagram
    participant User
    participant SourceChain as Source Chain
    participant Circle
    participant Guardians as Wormhole Guardians
    participant Relayers
    participant DestinationChain as Destination Chain

    User->>SourceChain: Initiate USDC transfer <br> with destination action
    SourceChain->>Circle: Burn USDC and package <br> transfer + action
    Circle->>Circle: Attest burn and issue signature
    Circle->>Guardians: Package message with <br> Circle signature
    Guardians->>Guardians: Verify & sign package
    Relayers->>Guardians: Pick up signed package
    Relayers->>DestinationChain: Relay package <br> automatically
    DestinationChain->>Circle: Verify Circle attestation
    DestinationChain->>DestinationChain: Mint native USDC
    DestinationChain->>DestinationChain: Execute destination action (optional)
```

!!! note 

>     For a cross-chain transfer to be successful, both the source and destination chains must be among those supported by [Circle's CCTP](https://developers.circle.com/stablecoins/supported-domains).

Reviewer 2: ```suggestion
    For a cross-chain transfer to be successful, both the source and destination chains must be among those supported by [Circle's CCTP](https://developers.circle.com/stablecoins/supported-domains){target=\_blank}.
```

Now that you're familiar with CCTP, here is a list of resources for more hands-on practice:

> - [**Get started with CCTP Bridge**](/docs/products/cctp-bridge/get-started/) - perform a multichain USDC transfer from Avalanche to Sepolia using Wormhole's TypeScript SDK and Circle's CCTP

Reviewer 2: Just FYI, in this case I would **not** add a target blank to this link because next steps is like ok where else on the docs can we send them. But I would add target blank to external links, such as circle's docs, so that we can try to keep users on our docs

- [**Complete USDC Transfer Flow**](Todo) - execute a USDC cross-chain transfer using Wormhole SDK and Circle's CCTP, covering manual, automatic, and partial transfer recovery

> - [**Checkout Circle's CCTP Docs**](https://developers.circle.com/stablecoins/cctp-getting-started) - learn more about Circle's cross-chain transfer protocol in their documentation

Reviewer 2: ```suggestion
- [**Checkout Circle's CCTP Docs**](https://developers.circle.com/stablecoins/cctp-getting-started){target=\_blank} - learn more about Circle's cross-chain transfer protocol in their documentation
```
